rust-runtime/aws-smithy-http-server/src/request/request_id.rs

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Request IDs
+//!
+//! RequestID is an element that uniquely identifies a client request. RequestID is used by services to map all logs, events and
+//! specific data to a single operation. This RFC discusses whether and how smithy-rs can make that value available to customers.
+//!
+//! Services use a RequestID to collect logs related to the same request and see its flow through the various operations,
+//! help clients debug requests by sharing this value and, in some cases, use this value to perform their business logic. RequestID is unique across a service at least within a certain timeframe.
+//!
+//! RequestIDs are not to be used by multiple services, but only within a single service.
+//!
+//! For client request IDs, the process will be, in order:
+//! * If a header is found matching one of the possible ones, use it
+//! * Otherwise, None
+//!
+//! Server request IDs are opaque and generated by the service itself.
+//!
+//! Your handler can now optionally take as input a [`ServerRequestId`] and an [`Option<ClientRequestId>`].
+//!
+//! ```rust,ignore
+//! pub async fn handler(
+//!     _input: Input,
+//!     server_request_id: ServerRequestId,
+//!     client_request_id: Option<ClientRequestId>,
+//! ) -> Output {
+//!     /* Use server_request_id and client_request_id */
+//!     todo!()
+//! }
+//!
+//! let app = Service::builder_without_plugins()
+//!     .operation(handler)
+//!     .build().unwrap();
+//!
+//! let app = app.layer(&ServerRequestIdProviderLayer::new()); /* Generate a server request ID */
+//! let app = app.layer(&ClientRequestIdProviderLayer::new(&["x-request-id"])); /* Provide your handler with the client request ID */
+//!
+//! let bind: std::net::SocketAddr = format!("{}:{}", args.address, args.port)
+//!     .parse()
+//!     .expect("unable to parse the server bind address and port");
+//! let server = hyper::Server::bind(&bind).serve(app.into_make_service());
+//! ```
```

Comment:
```suggestion
//! `aws-smithy-http-server` provides two types of request ids: [`ServerRequestId`] and [`ClientRequestId`].
//! 
//! ## `ServerRequestId`
//! 
//! A [`ServerRequestId`] is an opaque random identifier generated by the server every time it receives a request. 
//! It uniquely identifies the request within that service instance. It can be used to collate all logs, events and
//! data related to a single operation.
//! 
//! The [`ServerRequestId`] can be returned to the caller, who can in turn share the [`ServerRequestId`] to help the service owner in troubleshooting issues related to their usage of the service.
//!
//! The [`ServerRequestId`] is not meant to be propagated to downstream dependencies of the service. You should rely on a distributed tracing implementation for correlation purposes (e.g. OpenTelemetry). 
//!
//! ## `ClientRequestId`
//!
//! Clients can choose to provide their own identifier for the requests they are submitting.  
//! [`ClientRequestId`] can be used by the service to extract the client-provided request id - usually for logging purposes.
//!
//! We expect client request ids to be encoded as headers in the incoming request. The server will try to extract a value from the if a matching header is found.
//! Handlers and middlewares can use `Option<ClientRequestId>` to leverage the client request id - it will be set to `Some` if one was found, `None` otherwise.
//! 
//! ## Examples
//!
//! Your handler can now optionally take as input a [`ServerRequestId`] and an [`Option<ClientRequestId>`].
//!
//! ```rust
//! pub async fn handler(
//!     _input: Input,
//!     server_request_id: ServerRequestId,
//!     client_request_id: Option<ClientRequestId>,
//! ) -> Output {
//!     /* Use server_request_id and client_request_id */
//!     todo!()
//! }
//!
//! let app = Service::builder_without_plugins()
//!     .operation(handler)
//!     .build().unwrap();
//!
//! let app = app.layer(&ServerRequestIdProviderLayer::new()); /* Generate a server request ID */
//! let app = app.layer(&ClientRequestIdProviderLayer::new(&["x-request-id".into()])); /* Provide your handler with the client request ID */
//!
//! let bind: std::net::SocketAddr = format!("{}:{}", args.address, args.port)
//!     .parse()
//!     .expect("unable to parse the server bind address and port");
//! let server = hyper::Server::bind(&bind).serve(app.into_make_service());
//! ```
```